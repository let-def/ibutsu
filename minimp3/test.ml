open Minimp3

let t = make ()

let input =
  "\255\243\136\196\000\000\000\000\000\000\000\000\000Xing\000\000\000\015\
   \000\000\000\011\000\000\t$\000BBBBBBBBB\\\\\\\\\\\\\\\\\\vvvvvvvvv\137\
   \137\137\137\137\137\137\137\137\157\157\157\157\157\157\157\157\157\176\
   \176\176\176\176\176\176\176\176\193\193\193\193\193\193\193\193\193\209\
   \209\209\209\209\209\209\209\209\225\225\225\225\225\225\225\225\225\250\
   \250\250\250\250\250\250\250\250\250\250\250\250\250\250\250\250\250\000\
   \000\000,LAME3.82\000\000\000\000\000\000\000\000\000\000\000\000\000\000\
   \000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\
   \000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\
   \000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\
   \000\000\000\000\000\000\000\000BBBBBBBBB\\\\\\\\\\\\\\\\\\vvvvvvvvv\137\
   \137\137\137\137\137\137\137\137\157\157\157\157\157\157\157\157\157\176\
   \176\176\176\176\176\176\176\176\193\193\193\193\193\193\193\193\193\209\
   \209\209\209\209\209\209\209\255\243\152\196\000K\212\nD\001\158\176\000\
   \255\171\0180\129\015\003\024 \154\255q\028\211\015p\2030?-\143\248\025\
   \150\026%\024\001\129Hx\240\250\b.[\003\195\205\168\012\196\190P7\162v\
   \252\242h\018\192i\253\026\129\147\177L\006^\006\224\027\142-_'\212\\:O\
   \129\1534\030\006PB\152\025L\021@bd\181\129\151qk\250i\026\030>h\006GL\b\
   \025\031\024\192g8\140\001\164\194\204\006\015\001\168\015\000\031\217\205\
   \0159\163\001\135@\232\006\027\130p,\011\000\193\160l\003\018\193\248\006\
   \131 \024,\t_\244\141\015:\021 \192a\224G\001\135\000D\006:\005\144\024N\
   \003\160al\022\128\161\023\003\016B\128\0123\001@1\164\027\191\253>\239\
   \248\021\005\160\028\r\193\192\016\012\012\132\0160\142\005\192\195\024A\
   \003\000\224(\012\015\000A\021\003\001 <\006\128x\024/\003\255\255\255\255\
   \129 \020\002\128\016\006\0018\\\225\015\006\199\193 \000\002\12808\208m\
   \153$\005\128\024_\193>\001\129\160\r\255\255\255\255\254#\017\007\131cbD\
   \023\208;\001\128\199\178L\149\0250l\022\026\172w\131b\198\176\165\021\000\
   %d\203L\207M\020A\225\151\025\004\173\024B\177~\194(a\227\162\022\244\029`\
   \136\211pT\250nA\021\007p\016\132:\170\155\236\255\243h\196+8\187>\200\007\
   \153\192\001\181ZZ\023\135a*\146\228\182-\209:\221H\203%rmH\212\017\147\
   \218N\007\1382\232\210V\228\170\255i\250\214K\000QVZ\202!\011\230\024\131\
   &Zl!\195\145>\137\215\007\195\159\216\164m\251\138a\017uh\223g\250\207\183\
   \236\162\004N\136\235\247\025\183-\228\166\154j_;\021\148\216\165\199\157\
   \2393gqxn_\131\191\015\207\206s\234R\209\229\150\247\158\187Kr\1752\199\
   \153\1275\255\249\247\015\195\159\255\2473\169_Xe\141\158\254\191\029\239\
   \012q\222<\203\247\207\253~\181\135;\223\215\239\255\159c\154\203Y\215\149\
   *t\135\031\253\148\222\133F\248G~\221\173\205\175\142S\254\156\1368\000@\
   \237Y\021\188\002\002K\193\227n\002d\255\243x\196\0198+\242\216\007\204x\
   \001%\130\168H\131\169\243\004s$\234W6F9\147\240\154\155\245\r\138Wud?\220\
   \145\017\015t\130\176\195\031\234\240\191#\142\194\177\n0\197R\0160\005\
   \221\132\152\135\017\146\\D\232\195:M@\193\145x#Hk\209]V?\185)C\023\208\
   \215\0075\011F\2211BnR\197\195\197\203'\195TjB\196JKl\194\183\250\213\247<\
   J\193\158\241ej\127\ni\230\132\200\207\003\227zq\167\164\140o1#\rb@\187^\
   \159C\196|k\210\030\023\019\180\248\219\162\204(N1\027U\141I\216\214\172\
   \006Hm\140\204\186\134\217\168\145\031\2351_?\129&kl\186\140\2235\029[qm\
   \247\226\218>^_r\223>\147\195\137%a\255\233\029\195p.\181R\255?\006*E\208D\
   \192n\022\021#b\172\218\150\157nAl\146\174RX\252\178^\245SV\157\185\255\
   \243X\196!(\235\174\216\006\195\007M^\000\129\0297\130\003\148G\241\128#n\
   \005\214\029\006,hi\202~\227N\1818\017\017LOe\214\138\192\216*^\rD\146\176\
   \148\189['\140\019\018\190\130\165\251\201\201\203/\209\214u\255Mh\182\150\
   \215i\249k; \163&\217v;\249\195~\252V\018\177\179fXj]\193m\024{\179y\177\
   \253\006\187s\171\220\151\218\n.\025\243\255\254\148\218\252o\252\233\179\
   F\r\014\031\231\252f:=\020\154\021\183\136\134\218*\238\219\016\006i\232o\
   V\b\148\214\r`\139fY\215M\238r\218L0\215e\174\203\138\2493\nW\220\144\"\
   \255\243X\196$)\219\210\196\006\203\198\221\202\211\184\129.\024\012\173mN\
   \165F\1957\\\020\nF7L(Zq\181D\244\220?\018)\020S\213M\163;>\144GO]\167\136\
   \200\253\141\216\147\202\248\177b\192\143\140\189{3U\180\218\245\237f\147\
   \238\180\219\253k\227\243\158\199\1691,?\164V\193\251\163k\185\219\240\206\
   \157Vi\250el\153\195\252\223\1412+Y\1636J\164\200y\205waT\167Kl\167v\135\
   \223(|Zf]\237C{\146\139%%5\n\155\242 \000Bg\160X\016@\026R\243Q\213\160\
   \181\219\140\r\016\175K\r\203\216\170s\003Q\204+\011\012\255\243H\196#&\
   \218\222\188\014\211\006\189\192\020\b\140K\133S{\003A\236\184t^\029\131\
   \210\r\199a\240'\015W\015@\212A=\136\233\029\143P\168\189\148\196\231\227H\
   \156\205\2190\133g\242\138/\202\167\166\023^\231\150\031\147G\135\175\216f\
   \150\210\164\250\254\222|\165V\229\014\028;\170\1473$>|co\201o\244\143\128\
   \203\"-\159\202\015\179\190r\255g(\235\006\205\240\165\031\230\222\142l\
   \022g\1857\255\240(T\234\001\156\191\143b\148\135p\230\t\255\243X\196\n%#z\
   \172\022\203\006\209JA\150\184.R\205\144\181'\225\224u]\185d2\203\024\019^\
   \156\148\220\154\145Q\225\218\144\026\169\016\148\156\026\146A\176\236\218\
   \223\147\200c\179\190\1652;\197\r\175vD\138\0023\138\0024j$`\204\223\207*\
   \180\227\210R\194\150\206z\169\163\127u)II\148\150\189\165x\249)B\227\252 \
   \185\185mt\191\025\239>\2313m\184\176\229\223V\249\194[K\217\132\198A\197\
   \138c\133\186\180xz\140\206\184>\019>o\028\232&-\192\218\001\147s\011\228h\
   D\217\003\001@aJ\2190\149\217\175\232\156\205\216\205X\021\028\132W\255\
   \243H\196\028 2b\156\022\195\006\173\150\198\204V\018\197QV\204\211\184\
   \234\186\238\248\tX\0226'\022\189}H\151[v\133\229\149\250\\\210\182{T,\243\
   >f\225\225&z\158D\219\017\161\222\147\020R\145\154\234\230.D3;\171\160\216\
   D\184\198\232: Ie \138A%\167k\255+\190\237/u\128\243\213m\017\161\012\128\
   L\212\177P}\r\239nRB%\209\164\166\002\158\1392\184\155\167\005\188\177(\
   \180\174QA\n\151\193+\246]\221\227j\212\255\243H\196\030 k&\148\015X\024\
   \001V\237\150R\006h)\195\012F8hm\024\137\138\145\153\154ea\138d\174f\243\
   \133]K\233v\242\166h\176\205t\244\140\250\153\162\218D~M\209\019\006qKU\r\
   \164\206\197\244&*y\207)\219\247:y\161\018\149-_\144\161\030\218\216^\151\
   \018\193\236\226\b\205\023\182z\\\237\207&\150\025B\245\029vQ\172\2216\007\
   ,=\142<`P\255\249\136i\185\129\128\188b\183\248\128\0191\192\195\191\143\
   \248\025J`{\255\243h\196\0317\011j\024\001\157\160\0000\129\200\130b\165\
   \2401&\129\179\224(\004\000J\"\223\129\139\012\006\166\128b\144\r\028\006\
   LeJ\254\000\130B\000\192\224\192c\217\0004\224\012\011\255\224\002P\000\
   \141\001\139<n\022\130\002\229\006[\255\241B\134Y\001\225\128\201\144\012\
   \208\015\004\006\012\000\024\211\129t\127\255\240\248\000\198\156,\007$\
   \000!\128H\1601\163\128\206\172\000\147\000i\207\000q\"\t\254\175\255\224\
   \002\b$$B\192\221\016\203\128D\016\025S\192iU\002\016@fO\001\153(\000\132@\
   \145P\001\028\004\003\255R\170\255\255\224e\011\129\138\002\030\144^ d\145\
   \133\210\016\184\024\000\2269\018\192\180aI\r\176\198\163\028=\133\146\170\
   LAME3.82\170\170\170\170\170\255\243\024\196\r\000\000\003H\001\192\000\
   \000\170\170\170\170\170\170\170\170\170\170\170\170\170\170\170\170\170\
   \170\170\170\170\170\170"

let pcm =
  Bigarray.Array1.create
    Bigarray.int16_signed Bigarray.c_layout buffer_size

let average (pcm : samples) count =
  let sum = ref 0.0 in
  for i = 0 to count - 1 do
    sum := !sum +. float pcm.{i}
  done;
  !sum /. float count

let () =
  let frame = empty_frame () in
  frame.frame_bytes <- -1;
  let offset = ref 0 in
  while frame.frame_bytes <> 0 && !offset < String.length input do
    let samples =
      decode_string t input
        ~offset:!offset ~length:(String.length input - !offset) pcm frame
    in
    Printf.printf
      "samples = %d average = %f\n\
       { frame_bytes = %d; channels = %d; hz = %d; \
       layer = %d; bitrate_kbps = %d }\n"
      samples (average pcm samples)
      frame.frame_bytes frame.channels frame.hz frame.layer frame.bitrate_kbps;
    offset := !offset + frame.frame_bytes
  done
