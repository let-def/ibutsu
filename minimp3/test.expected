samples = 576 average = 0.000000
{ frame_bytes = 288; channels = 1; hz = 16000; layer = 3; bitrate_kbps = 64 }
samples = 576 average = 0.000000
{ frame_bytes = 360; channels = 1; hz = 16000; layer = 3; bitrate_kbps = 80 }
samples = 576 average = 246.579861
{ frame_bytes = 216; channels = 1; hz = 16000; layer = 3; bitrate_kbps = 48 }
samples = 576 average = -14.934028
{ frame_bytes = 252; channels = 1; hz = 16000; layer = 3; bitrate_kbps = 56 }
samples = 576 average = -211.571181
{ frame_bytes = 180; channels = 1; hz = 16000; layer = 3; bitrate_kbps = 40 }
samples = 576 average = 4.192708
{ frame_bytes = 180; channels = 1; hz = 16000; layer = 3; bitrate_kbps = 40 }
samples = 576 average = 2030.373264
{ frame_bytes = 144; channels = 1; hz = 16000; layer = 3; bitrate_kbps = 32 }
samples = 576 average = -2340.375000
{ frame_bytes = 180; channels = 1; hz = 16000; layer = 3; bitrate_kbps = 40 }
samples = 576 average = -1738.871528
{ frame_bytes = 144; channels = 1; hz = 16000; layer = 3; bitrate_kbps = 32 }
samples = 576 average = -1057.493056
{ frame_bytes = 144; channels = 1; hz = 16000; layer = 3; bitrate_kbps = 32 }
samples = 576 average = -497.923611
{ frame_bytes = 216; channels = 1; hz = 16000; layer = 3; bitrate_kbps = 48 }
samples = 576 average = 0.180556
{ frame_bytes = 36; channels = 1; hz = 16000; layer = 3; bitrate_kbps = 8 }
